updated_tictactoe.py

```diff
@@ -0,0 +1,146 @@
+import sys
+
+
+def format_user(element):
+    """
+    Return a string representation of a board position that can be logged to console.
+    """
+    if element is None:
+        return '_'
+    elif element:
+        return 'x'
+    else:
+        return 'o'
+
+
+def print_board(board):
+    """
+    Print the board to console.
+    """
+    for row in board:
+        for position in row:
+            print(f"{format_user(position)} ", end="")
+        print()
+
+
+def get_user_move():
+    """
+    Get the users move as an input from the console.  This method valids that the number
+    is between 1 - 9 or "q" (for quiting the game).
+    """
+    user_input = input("Please enter a position 1 through 9 or enter \"q\" to quit: ")
+
+    # Simply quit the program.
+    if user_input == 'q':
+        sys.exit()
+
+    # Validate the number is in range 1 - 9.
+    if user_input in [str(i) for i in range(1, 10)]:
```

Comment:
Could you explain this bit to me please. It says "i" is not defined

Reply:
The variable `i` is only scoped to the list comprehension. This means it can only be accessed within the brackets that it is defined. Here, we are creating a list of strings of the numbers 1 - 9.

What is the exact error you are getting?